threestudio/models/guidance/stable_diffusion_vsd_guidance.py

```diff
@@ -398,32 +410,63 @@ def get_latents(
             latents = self.encode_images(rgb_BCHW_512)
         return latents
 
-    def forward(
+    def __call__(
         self,
         rgb: Float[Tensor, "B H W C"],
-        text_embeddings: Float[Tensor, "BB 77 768"],
+        prompt_utils: PromptProcessorOutput,
+        elevation: Float[Tensor, "B"],
+        azimuth: Float[Tensor, "B"],
+        camera_distances: Float[Tensor, "B"],
         mvp_mtx: Float[Tensor, "B 4 4"],
+        c2w: Float[Tensor, "B 4 4"],
         rgb_as_latents=False,
+        **kwargs,
     ):
         batch_size = rgb.shape[0]
 
         rgb_BCHW = rgb.permute(0, 3, 1, 2)
         latents = self.get_latents(rgb_BCHW, rgb_as_latents=rgb_as_latents)
 
-        grad = self.compute_grad_vsd(latents, text_embeddings, mvp_mtx)
+        # view-dependent text embeddings
+        text_embeddings = prompt_utils.get_text_embeddings(
+            elevation,
+            azimuth,
+            camera_distances,
+            view_dependent_prompting=self.cfg.view_dependent_prompting,
+        )
+
+        # input text embeddings, view-independent
+        text_embeddings_inp = prompt_utils.get_text_embeddings(
+            elevation, azimuth, camera_distances, view_dependent_prompting=False
+        )
+
+        if self.cfg.camera_condition_type == "extrinsics":
+            camera_condition = c2w
+        elif self.cfg.camera_condition_type == "mvp":
+            camera_condition = mvp_mtx
+        else:
+            raise ValueError(
+                f"Unknown camera_condition_type {self.cfg.camera_condition_type}"
+            )
+
+        grad = self.compute_grad_vsd(
+            latents, text_embeddings, text_embeddings_inp, camera_condition
+        )
 
         grad = torch.nan_to_num(grad)
 
         # reparameterization trick
         # d(loss)/d(latents) = latents - target = latents - (latents - grad) = grad
         target = (latents - grad).detach()
-        vsd_loss = 0.5 * F.mse_loss(latents, target, reduction="sum") / batch_size
+        loss_vsd = 0.5 * F.mse_loss(latents, target, reduction="sum") / batch_size
 
-        lora_loss = self.train_lora(latents, text_embeddings, mvp_mtx)
+        loss_lora = self.train_lora(
+            latents, text_embeddings, text_embeddings_inp, camera_condition
+        )
 
```

Comment:
remove the useless text_embeddings in the parameters

Reply:
done